openlibrary/core/observations.py

```diff
@@ -1,26 +1,245 @@
 from openlibrary import accounts
+
 from . import cache
+from . import db
+
+
+@cache.memoize(engine="memcache", key="observations", expires=config.get('observation_cache_duration'))
+def get_observations():
+    """
+    Returns a dictionary of observations that are used to populate forms for patron feedback about a book.
+
+    Dictionary has the following structure:
+    {
+        'observations': [
+            {
+                'id': observation_types.id,
+                'label': observation_types.type,
+                'description': observation_types.description,
+                'multi_choice': observation_types.allow_multiple_values,
+                'values': [ observation_values.value list ]
+            }
+        ]
+    }
+
+    return: Dictionary of all possible observations that can be made about a book.
+    """
+    observations = Observations.get_observation_types_and_values()
+
+    observation_dict = {}
+
+    for o in observations:
+        type = o['type']
+
+        if type not in observation_dict:
+            observation_dict[type] = {
+                'id': o['type_id'],
+                'label': type,
+                'description': o['description'],
+                'multi_choice': o['multi_choice'],
+                'values': []
+            }
+
+        observation_dict[type]['values'].insert(0, (o['value_id'], o['value'], o['prev_value']))
+
+    observation_list = []
 
-# URL for TheBestBookOn
-TBBO_URL = config.get('tbbo_url')
+    for v in observation_dict.values():
+        v['values'] = _sort_values(v['values'])
+        observation_list.append(v)
 
-def post_observation(data, s3_keys):
-    headers = {
-        'x-s3-access': s3_keys['access'],
-        'x-s3-secret': s3_keys['secret']
+    response = {
+        'observations': observation_list
     }
 
-    response = requests.post(TBBO_URL + '/api/observations', data=data, headers=headers)
+    return response
+
+def _sort_values(values_list):
+    """
+    Given a list of value tuples, returns a sorted list of observation values.
+    
+    Value tuples have the form (value ID, value, previous value ID).  Previous
+    value ID is the ID of the previous value in the list, or None if the value
+    is the first item in the list.
+
+    return: A sorted list of values.
+    """
+    # Add middle list item to sorted list
+    middle_item = values_list.pop(int(len(values_list) / 2))
+    sorted_list = [ middle_item[1] ]
+
+    # Previous id:
+    prev = middle_item[2]
+
+    # Next id:
+    next = middle_item[0]
+
+    map = {}
+    for i in values_list:
+        map[i[0]] = i
+
+    while len(map):
+        for k in list(map):
+            if map[k][0] == prev:
+                sorted_list.insert(0, map[k][1])
+                prev = map[k][2]
+                del map[k]
+            elif map[k][2] == next:
+                
+                sorted_list.append(map[k][1])
+                next = map[k][0]
+                del map[k]
```

Comment:
```suggestion
    values_dict = {i[0]: i for i in values_list}
    for key in list(values_dict):  # values_dict will change during iteration!
        value = values_dict
        if value[0] == prev:
            sorted_list.insert(0, value[1])
            prev = value[2]
            del values_dict[k]
        elif value[2] == next:        
            sorted_list.append(value[1])
            next = value[0]
            del values_dict[k]
```
Dict comprehension, don't shadow built-in, speed up access, document assumptions.

Reply:
The outer while loop is necessary.  `values_dict` can be understood as a dictionary of singly linked list nodes.  With each iteration, we are adding nodes to their proper places in the list.  Once the dictionary is empty, the list is complete and ordered correctly.

I'll add some comments to clarify what is happening with this function.

Also want to point out that there's probably a more efficient way of doing this, but this is only being called by `get_observations()` when the cached value expires (once per day currently, but could really be longer).